File: scripts/diagnostics/andromeda_catalog_sync.py
"""Compare the complete saved Egypt catalog and persist explicit source identities."""
import argparse
import hashlib
import json
from pathlib import Path
import re
import unicodedata
from andromeda_hotel_candidates import save

CATALOG_SHA = '01030bb9e23e0c87f8bed7c50628c8f56243b89f3e55a24151430766c1576641'


def names(value):
    # Parenthesized EX names are alternate names, not an excuse to remove BEACH/PALACE.
    value = unicodedata.normalize('NFKC', value).casefold().replace('ё','е').replace('&',' and ')
    parts = re.split(r'\(\s*(?:ex[.\s:-]*|быв[.\s:-]*)', value)
    result = set()
    for part in parts:
        words = re.findall(r'[^\W_]+', part.split(')')[0], re.UNICODE)
        words = [w for w in words if w not in {'hotel','hotels','resort','resorts','spa','the','and','отель'}]
        if len(''.join(words)) >= 5:
            result.add(' '.join(words))
    return result


def match(catalog, local, reviewed):
    if catalog['params'] != {'TOWNFROMINC':1,'STATEINC':3} or len(catalog['payload']['HOTELS']) != 905:
        raise ValueError('unexpected source catalog')
    if local.get('complete') is not True or local.get('country_id') != 1:
        raise ValueError('incomplete local catalog')
    hotels={int(h['id']):h for h in local['hotels']}
    aliases={i:set(names(h['name'])) for i,h in hotels.items()}
    for a in local['aliases']:
        aliases[int(a['hotel_id'])].update(names(a['alias']))
    index={}
    for identifier,values in aliases.items():
        for name in values:index.setdefault(name,set()).add(identifier)
    approved={r['external_hotel_id']:r['proposed_catalog_hotel_id'] for r in reviewed['rows'] if r['decision_status']=='proposed'}
    rows=[]
    for source in catalog['payload']['HOTELS']:
        external=str(source['id']); status='pending'; target=None; reason='no_unique_name'
        candidates=set()
        for name in (source['name'],source.get('lName','')):
            for key in names(name):candidates.update(index.get(key,set()))
        if str(source['stateKey'])!='3' or external=='2000073714':
            status='conflict';reason='country_or_known_catalog_offer_conflict'
        elif external in approved:
            target=approved[external]
            if target not in hotels:raise ValueError('reviewed local hotel disappeared')
            status='accepted';reason='owner_authorized_reviewed_identity'
        elif len(candidates)==1:
            target=next(iter(candidates));status='accepted';reason='unique_country_name_or_former_name'
        evidence={'source':source,'candidate_ids':sorted(candidates),'target_name':hotels[target]['name'] if target else None,'reason':reason}
        rows.append({'external_hotel_id':external,'local_hotel_id':target,'decision_status':status,
                     'evidence_json':json.dumps(evidence,ensure_ascii=False,sort_keys=True,separators=(',',':'))})
    return {'operation':'import','catalog_sha256':CATALOG_SHA,'rows':rows}


def main():
    parser=argparse.ArgumentParser();parser.add_argument('phase',choices=['prepare','apply'])
    parser.add_argument('--directory',type=Path,required=True);args=parser.parse_args();root=args.directory
    import anex_search3_owner_decisions as owner
    php=Path(__file__).with_suffix('.php').read_text().removeprefix('<?php')
    state_path=root/'sync-state.json'
    if args.phase=='prepare':
        if state_path.exists():raise ValueError('existing sync state; no replay')
        save(state_path,{'state':'export_inflight'},exclusive=True)
        local=owner.ssh_php(php,{'operation':'export'},maximum_bytes=4000000)
        if local.get('status')!='ok':raise ValueError('local export failed')
        save(root/'local-catalog.json',local,exclusive=True)
        raw=(root/'all.json').read_bytes()
        if hashlib.sha256(raw).hexdigest()!=CATALOG_SHA:raise ValueError('source digest changed')
        review_path=Path(__file__).resolve().parents[2]/'reports/andromeda-hotel-shortlist-20260909.json'
        request=match(json.loads(raw),local,json.loads(review_path.read_bytes()))
        save(root/'import-request.json',request,exclusive=True)
        save(state_path,{'state':'prepared','request_sha256':hashlib.sha256((root/'import-request.json').read_bytes()).hexdigest()})
        print(json.dumps({'local_hotels':len(local['hotels']),'source_hotels':905,'counts':{s:sum(r['decision_status']==s for r in request['rows']) for s in ['accepted','pending','conflict']}}))
    else:
        state=json.loads(state_path.read_bytes());raw=(root/'import-request.json').read_bytes()
        if state['state']!='prepared' or hashlib.sha256(raw).hexdigest()!=state['request_sha256']:raise ValueError('sync not prepared')
        save(state_path,{**state,'state':'import_inflight'})
        request=json.loads(raw);result=owner.ssh_php(php,request,maximum_bytes=4000000)
        save(root/'import-result.json',result,exclusive=True)
        actual={r['external_hotel_id']:(str(r['local_hotel_id']),r['decision_status'],r['evidence_sha256']) for r in result.get('rows',[])}
        expected={r['external_hotel_id']:(str(r['local_hotel_id']),r['decision_status'],hashlib.sha256(r['evidence_json'].encode()).hexdigest()) for r in request['rows']}
        # PDO numeric IDs may arrive as strings; normalize null identically above.
        if result.get('status')!='imported' or actual!=expected:raise ValueError('import readback mismatch; preserve state')
        save(state_path,{**state,'state':'completed','counts':result['counts'],'readback_verified':True})
        print(json.dumps({'state':'completed','counts':result['counts'],'inserted':result['inserted'],'readback_verified':True}))


if __name__=='__main__':main()

File: scripts/diagnostics/andromeda_catalog_sync.php
<?php
// Confined AnyTour catalog read / append-only Andromeda identity import.
error_reporting(0);
ob_start();
$pdo = null;
try {
    $root = realpath(getcwd());
    if (!$root || basename($root) !== 'anytoour.ru') throw new RuntimeException();
    require_once $root . (is_file($root.'/data/db-v1.php') ? '/data/db-v1.php' : '/v2/data/db-v1.php');
    $pdo = v2_data_db();
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    $raw = file_get_contents('php://stdin', false, null, 0, 4000001);
    if (strlen($raw)>4000000) throw new RuntimeException();
    $request = json_decode($raw, true, 32, JSON_THROW_ON_ERROR);
    if ($request['operation'] === 'export') {
        $pdo->exec('START TRANSACTION READ ONLY');
        $hotels = $pdo->query('SELECT id,name,country_id,country_name,region_name,subregion_name,category FROM catalog_hotels WHERE country_id=1 AND is_active=1 ORDER BY id LIMIT 10001')->fetchAll(PDO::FETCH_ASSOC);
        $aliases = $pdo->query('SELECT a.hotel_id,a.alias FROM hotel_aliases a JOIN catalog_hotels h ON h.id=a.hotel_id WHERE h.country_id=1 AND h.is_active=1 ORDER BY a.hotel_id,a.alias LIMIT 30001')->fetchAll(PDO::FETCH_ASSOC);
        if (count($hotels)>10000 || count($aliases)>30000) throw new RuntimeException();
        $pdo->exec('ROLLBACK');
        $result = ['status'=>'ok','hotels'=>$hotels,'aliases'=>$aliases,'country_id'=>1,'complete'=>true];
    } elseif ($request['operation'] === 'import') {
        $rows=$request['rows'];
        if (!is_array($rows) || count($rows)!==905 || !preg_match('/^[a-f0-9]{64}$/D',$request['catalog_sha256']??'')) throw new RuntimeException();
        $seen=[];
        foreach ($rows as $r) {
            $id=$r['external_hotel_id']; $target=$r['local_hotel_id'];
            if (!is_string($id) || !preg_match('/^[1-9][0-9]{0,31}$/D',$id) || isset($seen[$id])
                || !in_array($r['decision_status'],['accepted','pending','conflict'],true)
                || ($r['decision_status']==='accepted' ? (!is_int($target)||$target<1||$id==='2000073714') : $target!==null)
                || !is_string($r['evidence_json']) || strlen($r['evidence_json'])>20000) throw new RuntimeException();
            $seen[$id]=true;
        }
        $pdo->exec("CREATE TABLE IF NOT EXISTS andromeda_hotel_identities (supplier_namespace VARCHAR(40) NOT NULL,external_hotel_id VARCHAR(32) NOT NULL,local_hotel_id INT NULL,decision_status VARCHAR(16) NOT NULL,catalog_sha256 CHAR(64) NOT NULL,evidence_sha256 CHAR(64) NOT NULL,evidence_json MEDIUMTEXT NOT NULL,created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,PRIMARY KEY(supplier_namespace,external_hotel_id),KEY(local_hotel_id)) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4");
        $pdo->beginTransaction();
        $existing=$pdo->query("SELECT * FROM andromeda_hotel_identities WHERE supplier_namespace='andromeda_catalog' FOR UPDATE")->fetchAll(PDO::FETCH_ASSOC);
        $index=[]; foreach($existing as $r)$index[$r['external_hotel_id']]=$r;
        $lookup=$pdo->prepare('SELECT id,name FROM catalog_hotels WHERE id=? AND country_id=1 AND is_active=1');
        $insert=$pdo->prepare("INSERT INTO andromeda_hotel_identities(supplier_namespace,external_hotel_id,local_hotel_id,decision_status,catalog_sha256,evidence_sha256,evidence_json) VALUES('andromeda_catalog',?,?,?,?,?,?)");
        $inserted=0;
        foreach($rows as $r){
            $sha=hash('sha256',$r['evidence_json']);
            if(isset($index[$r['external_hotel_id']])){
                $old=$index[$r['external_hotel_id']];
                if($old['evidence_sha256']!==$sha || $old['decision_status']!==$r['decision_status'] || (string)$old['local_hotel_id']!==(string)$r['local_hotel_id']) throw new RuntimeException();
                continue;
            }
            if($r['local_hotel_id']!==null){
                $lookup->execute([$r['local_hotel_id']]); $hotel=$lookup->fetch(PDO::FETCH_ASSOC);
                $evidence=json_decode($r['evidence_json'],true,32,JSON_THROW_ON_ERROR);
                if(!$hotel || $hotel['name']!==$evidence['target_name']) throw new RuntimeException();
            }
            $insert->execute([$r['external_hotel_id'],$r['local_hotel_id'],$r['decision_status'],$request['catalog_sha256'],$sha,$r['evidence_json']]);++$inserted;
        }
        $counts=$pdo->query("SELECT decision_status,COUNT(*) AS total FROM andromeda_hotel_identities WHERE supplier_namespace='andromeda_catalog' GROUP BY decision_status")->fetchAll(PDO::FETCH_KEY_PAIR);
        $pdo->commit();
        $readback=$pdo->query("SELECT external_hotel_id,local_hotel_id,decision_status,evidence_sha256 FROM andromeda_hotel_identities WHERE supplier_namespace='andromeda_catalog' ORDER BY external_hotel_id")->fetchAll(PDO::FETCH_ASSOC);
        $result=['status'=>'imported','inserted'=>$inserted,'preserved'=>count($existing),'counts'=>$counts,'rows'=>$readback];
    } else throw new RuntimeException();
} catch(Throwable $error) {
    if($pdo instanceof PDO && $pdo->inTransaction())$pdo->rollBack();
    $result=['status'=>'failed'];
}
while(ob_get_level())ob_end_clean();
$out=json_encode($result,JSON_UNESCAPED_UNICODE|JSON_UNESCAPED_SLASHES);
echo strlen($out)<=4000000?$out:'{"status":"failed"}';
